File: 0020-valid-parentheses/0020-valid-parentheses.py
## space: o(n)
## time: o(n)

class Solution:
    def isValid(self, s: str) -> bool:
        
        ## create an empty stack
        stack = []
        ## map the closing brackets to open brackets using hashmaps
        
        closeToOpen = { ")" : "(" , "]" : "[" , "}" : "{"}
        
        ##iterate through the input string and check if the key(closing bracket in the stack) -1 is the last element of stack
        
        for c in s:
            if c in closeToOpen:
                
                if stack and stack[-1]  == closeToOpen[c]:
                    ## start popping
                    stack.pop()
                    
                else:
                    ## means there is no matching 
                    return False
                
            else:
                ## this else case inidcates the brakcet is open and hence we add it to the stack stack follows LIFO order
                stack.append(c)
                
        if not stack:
            return True
        else:
            return False
            
  # Create an instance of the Solution class
##solution = Solution()

# Test cases
##input_str1 = "()"
##input_str2 = "({[()]})"

# Check if the input strings have valid brackets
##print(solution.isValid(input_str1))  # Should print True
##print(solution.isValid(input_str2))  # Should print True
##      

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        
        stack<char> stack;
        
        unordered_map<char,char> closeToOpen = {{')', '('}, {']', '['}, {'}', '{'}};
        
        for(char c : s){
            if(closeToOpen.find(c) != closeToOpen.end()){
                if(!stack.empty() && stack.top() == closeToOpen[c]){
                    stack.pop();
                }
                else{
                    return false;
                }
            }
            else{
                stack.push(c);
            }
        }
        return stack.empty();
        
    }
};

/*
int main() {
    Solution solution;

    // Test cases
    std::string input_str1 = "()";
    std::string input_str2 = "({[()]})";

    // Check if the input strings have valid brackets
    std::cout << std::boolalpha;  // Print "true" or "false" instead of 1 or 0
    std::cout << solution.isValid(input_str1) << std::endl;  // Should print true
    std::cout << solution.isValid(input_str2) << std::endl;  // Should print true

    return 0;
}
*/